libraries/chain/database.cpp

```diff
@@ -2005,6 +2014,94 @@ void database::apply_operation(const operation& op)
    notify_post_apply_operation( note );
 }
 
+void database::process_tx_bandwidth(const signed_transaction& trx) {
+   account_bandwidth_object trx_bandwidth_data;
+
+   trx_bandwidth_data.total_bandwidth = fc::raw::pack_size(trx);
+   trx_bandwidth_data.total_ops_count = trx.operations.size();
+
+   bool fee_free_op_present = false;
+   bool paid_op_present     = false;
+   for (const operation& op : trx.operations) {
+      if (is_fee_free_operation(op) == true) {
+         // Count standalone fee-free operations bandwidth
+         trx_bandwidth_data.act_fee_free_bandwidth += fc::raw::pack_size(op);
+         trx_bandwidth_data.act_fee_free_ops_count++;
+         fee_free_op_present = true;
+      }
+      else {
+         paid_op_present = true;
+      }
+   }
+
+   // If there are only fee-free operations, count as consumed fee-free bandwidth also transaction meta info,
+   // if there are also paid operations, count as consumed fee-free bandwidth ONLY standalone operations(see above)
+   if (paid_op_present == false && fee_free_op_present == true) {
+      trx_bandwidth_data.act_fee_free_bandwidth = trx_bandwidth_data.total_bandwidth;
```

Comment:
Does this logic means that with 1 paid operation one can pack the rest 8k of the transaction with free ops?

Reply:
this logic means: if there is no paid operation, count as fee-free bandwidth also transaction meta info, like expiration, ref_block, etc... If there is at least 1 paid operation, count as fee-free bawidth only operation data